src/ds/flaglock.h

```diff
@@ -6,21 +6,144 @@
+     */
+    std::atomic_flag flag = ATOMIC_FLAG_INIT;
+
+    constexpr DebugFlagWord() = default;
+
+    template<typename... Args>
+    constexpr DebugFlagWord(Args&&... args) : flag(std::forward<Args>(args)...)
+    {}
+
+    /**
+     * @brief set_owner
+     * Record the identity of the locker.
+     */
+    void set_owner()
+    {
+      SNMALLOC_ASSERT(nullptr == owner);
+      owner = get_thread_identity();
+    }
+
+    /**
+     * @brief clear_owner
+     * Set the identity to null.
+     */
+    void clear_owner()
+    {
+      SNMALLOC_ASSERT(get_thread_identity() == owner);
+      owner = nullptr;
+    }
+
+    /**
+     * @brief assert_not_owned_by_current_thread
+     * Assert the lock should not be held already by current thread.
+     */
+    void assert_not_owned_by_current_thread()
+    {
+      SNMALLOC_ASSERT(get_thread_identity() != owner);
+    }
+
+  private:
+    using ThreadIdentity = int const*;
+
+    /**
+     * @brief owner
+     * We use a pointer to TLS field as the thread identity.
+     * std::thread::id can be another solution but it does not
+     * support `constexpr` initialisation on some platforms.
+     */
+    ThreadIdentity owner = nullptr;
+
+    /**
+     * @brief get_thread_identity
+     * @return The identity of current thread.
+     */
+    inline ThreadIdentity get_thread_identity()
+    {
+      static thread_local int SNMALLOC_THREAD_IDENTITY = 0;
+      return &SNMALLOC_THREAD_IDENTITY;
+    }
+  };
+
+  /**
+   * @brief The ReleaseFlagWord struct
+   * The shares the same structure with DebugFlagWord but
+   * all member functions associated with ownership checkings
+   * are empty so that they can be optimised out at Release mode.
+   */
+  struct ReleaseFlagWord
+  {
+    std::atomic_flag flag = ATOMIC_FLAG_INIT;
+
+    constexpr ReleaseFlagWord() = default;
+
+    template<typename... Args>
+    constexpr ReleaseFlagWord(Args&&... args)
+    : flag(std::forward<Args>(args)...)
+    {}
+
+    void set_owner() {}
+    void clear_owner() {}
+    void assert_not_owned_by_current_thread() {}
+  };
+
+#ifdef NDEBUG
+  using FlagWord = ReleaseFlagWord;
+#else
+  using FlagWord = DebugFlagWord;
+#endif
+
   class FlagLock
   {
   private:
-    std::atomic_flag& lock;
+    FlagWord& lock;
 
   public:
-    FlagLock(std::atomic_flag& lock) : lock(lock)
+    FlagLock(FlagWord& lock) : lock(lock)
     {
-      while (lock.test_and_set(std::memory_order_acquire))
+#ifdef __cpp_lib_atomic_flag_test
+      do
+      {
+        // test once with acquire ordering at the beginning
+        if (!lock.flag.test_and_set(std::memory_order_acquire))
+        {
+          break;
+        }
+        // assert_not_owned_by_current_thread is only called when the first
+        // acquiring is failed; which means the lock is already held somewhere
+        // else.
+        lock.assert_not_owned_by_current_thread();
+        // only need to do relaxed ordering in the busy spin loop
```

Comment:
Is this true?  If I understand the spec correctly (and, to be clear, I definitely do not fully understand the C++11 atomics spec), the relaxed read does not establish a happens-before relationship with the release store.  This means that the spin loop is allowed to spin forever (and a compiler is free to transform it into a single load followed by an infinite loop).

Reply:
You are definitely write. x86_64 is generating `mov` anyway so I wrongly placed relaxed ordering here. But it is because x86_64's mov itself has a lot of restrictions. I re-checked the standard:

> **memory_order_relaxed** Relaxed operation: there are no synchronization or ordering  constraints imposed on other reads or writes, only this operation's  atomicity is guaranteed

> **memory_order_acquire** A load operation with this memory order performs the acquire operation on the affected memory location: no reads or writes in the current thread can be reordered before this load. All writes in other threads that release the same atomic variable are visible in the current thread.

So yes, the ordering here should still be `acquire` but this optimisation will also work since there will be no `xchg` while spinning.

Reply:
A relatex discussion from the following link:
https://rigtorp.se/spinlock/

Reply:
I am not convinced anything in the standard actually makes progress guarantees.  I was fairly convinced you couldn't optimise away a relaxed atomic, but there are lots of re-orderings that can occur.  Anyway, I am happy with the current slightly more pessimistic code.  And this should not be on a performance critical path, so is fine to be less than optimal.